Fixed Type initial PR

VoR0220 opened the pull request:
Fixes https://github.com/ethereum/solidity/issues/428

Major changes:
- convert constant integer to rationals (two big ints)
- lay the foundations for the fixed type
- some changes to the operators
- numerous tests added

Commit:
initial work for fixed types...potentially needing a constant literal type for this

notation

Rational implemented...trying to figure out exponential

fix for token bug, also quick fix for the wei and seconds

fixed problem with var...probably a conversion problem for fixed in size capabilities

adding fixed type tests

Removing bitshift and regrouping fixed type tests together

size capabilities functioning properly for fixed types

got exponents up and working with their inverse, changed a few of the tests....something is working that likely shouldn't be

slight changes to how to flip the rational negative around...still trying to figure it out

tests added

updated tests

odd differences in trying soltest from solc binary, let me know if you can replicate

test not working for odd reason

fixed test problem with fixed literals...still need a way to log this error

broken up the tests, added some, changed some things in types and began compiler work

moar tests and prepping for rebuilding much of the types.cpp file

further fixing

initial work for fixed types...potentially needing a constant literal type for this

chriseth commented on libsolidity/ast/Types.cpp:
```diff
@@ -251,17 +260,30 @@ IntegerType::IntegerType(int _bits, IntegerType::Modifier _modifier):
+		else if (isSigned())
+			return convertTo.isSigned();
+		else
+			return !convertTo.isSigned() || convertTo.leftBits() > m_bits;
```
Why allow the second part?

chriseth commented on libsolidity/ast/Types.cpp:
```diff
@@ -339,29 +361,159 @@ MemberList::MemberMap IntegerType::nativeMembers(ContractDefinition const*) cons
 }
 
-bool IntegerConstantType::isValidLiteral(const Literal& _literal)
+FixedType::FixedType(int _Lbits, int _Rbits, FixedType::Modifier _modifier):
```
Please use the terms `integer bits` and `fractional bits`.

chriseth commented on libsolidity/ast/Types.cpp:
```diff
@@ -339,29 +361,159 @@ MemberList::MemberMap IntegerType::nativeMembers(ContractDefinition const*) cons
 	catch (...)
 	{
-		return false;
+			return false;
```
Indentation

chriseth commented on libsolidity/ast/Types.cpp:
```diff
@@ -339,29 +361,159 @@ MemberList::MemberMap IntegerType::nativeMembers(ContractDefinition const*) cons
-		bigint x(_literal.value());
+		string literal = _literal.value();
+		auto literalEnd = remove(literal.begin(), literal.end(), '.');
+		bigint x(string(literal.begin(), literalEnd));
```
What does this check? Please add a comment about which literals are valid.

VoR0220 replied:
Definitely should have written that in. This checks to see if it is a valid literal or not. The remove function is to make sure that the radix is removed, otherwise there will be an error thrown when trying to convert to a bigint. Essentially what this is checking is that there are no alphabetical characters, and that there are no odd symbols in there that would not be considered part of an integer.

chriseth commented on libsolidity/ast/Types.cpp:
```diff
@@ -339,29 +361,159 @@ MemberList::MemberMap IntegerType::nativeMembers(ContractDefinition const*) cons
+	{
+		m_modifier = Modifier::Fixed;
+		auto literalEnd = remove(literal.begin(), literal.end(), '.');
+		m_scalingFactor = distance(radix, literalEnd);
```
Not sure if we should use decimal or binary here

VoR0220 replied:
Just from the wikipedia page you gave me, apparently binary will be faster, however, decimal will be more accurate and generally easier to code and understand. My vote goes to decimal.

chriseth commented on libsolidity/ast/Types.cpp:
```diff
@@ -389,43 +541,100 @@ IntegerConstantType::IntegerConstantType(Literal const& _literal)
 			return true;
-		return false;
+		int forSignBit = (targetType->isSigned() ? 1 : 0);
+		if (isInteger())
```
This seems really complicated. What are the exact rules?

VoR0220 replied:
This case is for if our current constantNumber is an integer or not, the else is if it's a fixed type. If it's an integer, we only focus on the bits left of the radix. If it's a fixed type, we make sure that the size of the integer bits and the size of the fractional bits for the type we are converting to are greater or equal to what we currently have so that they fit, respectively. We have the sign bit for obvious reasons in there.

chriseth commented on libsolidity/ast/Types.cpp:
```diff
@@ -451,44 +660,93 @@ TypePointer IntegerConstantType::binaryOperatorResult(Token::Value _operator, Ty
-			value = m_value + other.m_value;
+			if (modifier == Modifier::Fixed)
+			{
+				bigint leftSide = leftOfRadix() + other.leftOfRadix();
```
This looks like we actually do not need a special case for fixed point arithmetics.

VoR0220 replied:
I think you're right. This special case should be the default case for all ConstantNumbers.

chriseth commented on libsolidity/ast/Types.h:
```diff
@@ -132,7 +132,7 @@ class Type: private boost::noncopyable, public std::enable_shared_from_this<Type
 	enum class Category
 	{
-		Integer, IntegerConstant, StringLiteral, Bool, Real, Array,
+		Integer, NumberConstant, StringLiteral, Bool, Fixed, Array,
```
I think we should call it `FixedPoint`, otherwise it might be mistaken for `FixedBytes`.

VoR0220 replied:
That's fair.

chriseth commented on libsolidity/ast/Types.h:
```diff
@@ -308,20 +308,67 @@ class IntegerType: public Type
- * Integer constants either literals or computed. Example expressions: 2, 2+10, ~10.
+ * A fixed point type number (signed, unsigned).
+ */
+class FixedType: public Type
```
Also here, I think `FixedPointType` would be better.

VoR0220 replied:
agreed

chriseth commented on libsolidity/ast/Types.h:
```diff
@@ -308,20 +308,67 @@ class IntegerType: public Type
-	virtual Category category() const override { return Category::IntegerConstant; }
+	enum class Modifier
+	{
+		Integer, Fixed
```
Do we really need this? I would say `1.0` is an integer, as is the result of `(2.5 * 2)`.

VoR0220 replied:
mmmmmmm....I suppose we could do that for those cases, but I think whenever melding between the two of them it helps in distinguishing them. I could be wrong however. But that's how I see it currently. There's some rules I've baked in that require this distinction.

chriseth commented on libsolidity/ast/Types.h:
```diff
@@ -338,9 +385,26 @@ class IntegerConstantType: public Type
 private:
+
 	bigint m_value;
+	short m_scalingFactor;
```
Please add a comment, especially about what happens if we e.g. compute `1.0 / 3`.

VoR0220 replied:
certainly.

chriseth commented on libsolidity/formal/Why3Translator.cpp:
```diff
@@ -428,7 +428,7 @@ bool Why3Translator::visit(BinaryOperation const& _binaryOperation)
 	Token::Value const c_op = _binaryOperation.getOperator();
 
-	if (commonType.category() == Type::Category::IntegerConstant)
+	if (commonType.category() == Type::Category::NumberConstant)
```
Please throw an error if this is not an integer (not sure if this is the correct place).

VoR0220 replied:
will write that in.

VoR0220 replied:
I'm actually not certain how to go about this one. I'm leaving it for now...going to focus on the other parts of this.

chriseth replied:
Check that denominator is 1, otherwise call `error(_identifier, "Fractional numbers not supported.")`.

VoR0220 replied:
done

Commit:
Rational implemented...trying to figure out exponential

Commit:
fix for token bug, also quick fix for the wei and seconds

Commit:
fixed problem with var...probably a conversion problem for fixed in size capabilities

adding fixed type tests

Removing bitshift and regrouping fixed type tests together

chriseth commented:
Please rebase (PR contains merge commits and github has no way to see how potential merge conflicts were resolved).

Commit:
size capabilities functioning properly for fixed types

Commit:
got exponents up and working with their inverse, changed a few of the tests....something is working that likely shouldn't be

slight changes to how to flip the rational negative around...still trying to figure it out

tests added

updated tests

odd differences in trying soltest from solc binary, let me know if you can replicate

test not working for odd reason

fixed test problem with fixed literals...still need a way to log this error

broken up the tests, added some, changed some things in types and began compiler work

moar tests and prepping for rebuilding much of the types.cpp file

further fixing

infinite loop still happening but it's somewhere in the fixedPoint methodd

fractional bits needed algo improved! Eliminated 2 errors

Corrected problems with the previous commit. No infinite loops. Actually appear to have corrected an error

chriseth commented on libsolidity/ast/Types.cpp:
```diff
@@ -340,11 +368,138 @@ MemberList::MemberMap IntegerType::nativeMembers(ContractDefinition const*) cons
+		m_integerBits + m_fractionalBits <= 256 && 
+		m_integerBits % 8 == 0 && 
+		m_fractionalBits % 8 == 0,
+		"Invalid bit number(s) for fixed type: " + 
```
I think we have to add another restriction here: If the fixed point type is signed, the integer part cannot have zero bits, because we need at least one bit for the sign.
An assert should be added here, and Token.cpp should not generate/parse the tokens `fixed0x8`, ..., `fixed0x256` in the first place.

VoR0220 replied:
fair point

VoR0220 replied:
dropped as can add sign bit to fractional side.

VoR0220 replied:
skipping because we can assign sign bit to the fractional side.

chriseth commented on libsolidity/ast/Types.cpp:
```diff
@@ -340,11 +368,138 @@ MemberList::MemberMap IntegerType::nativeMembers(ContractDefinition const*) cons
+		else if (isSigned())
+			return convertTo.isSigned();
+		else
+			return !convertTo.isSigned() || (convertTo.m_integerBits > m_integerBits && convertTo.m_fractionalBits > m_fractionalBits);
```
We can drop the condition about fractional bits, I think.

VoR0220 replied:
why?

chriseth replied:
Because `convertTo.m_fractionalBits == m_fractionalBits` is also fine, and we already checked `<` above.

chriseth commented on libsolidity/ast/Types.cpp:
```diff
@@ -340,11 +368,138 @@ MemberList::MemberMap IntegerType::nativeMembers(ContractDefinition const*) cons
+	// All fixed types can be compared
+	if (Token::isCompareOp(_operator))
+		return commonType;
+	if (Token::isBooleanOp(_operator))
```
I think we should not allow boolean operators for fixed point types.

chriseth commented on libsolidity/ast/Types.cpp:
```diff
@@ -340,11 +368,138 @@ MemberList::MemberMap IntegerType::nativeMembers(ContractDefinition const*) cons
+	return commonType;
+}
+
+bool ConstantNumberType::isValidLiteral(const Literal& _literal)
```
`Literal const&`

VoR0220 replied:
on it

chriseth commented on libsolidity/ast/Types.cpp:
```diff
@@ -340,11 +368,138 @@ MemberList::MemberMap IntegerType::nativeMembers(ContractDefinition const*) cons
 	{
-		bigint x(_literal.value());
+		rational numerator;
+		rational denominator = bigint(1);
```
`rational denominator(1)` does not work?

chriseth commented on libsolidity/ast/Types.cpp:
```diff
@@ -340,11 +368,138 @@ MemberList::MemberMap IntegerType::nativeMembers(ContractDefinition const*) cons
+		else
+		{
+			numerator = bigint(_literal.value());
+			rational x = numerator;
```
Is this needed?

VoR0220 replied:
no it is not. Good catch.

chriseth commented on libsolidity/ast/Types.cpp:
```diff
@@ -340,11 +368,138 @@ MemberList::MemberMap IntegerType::nativeMembers(ContractDefinition const*) cons
+		auto radixPoint = find(_literal.value().begin(), _literal.value().end(), '.');		
+		if (radixPoint != _literal.value().end())
+		{
+			//problem here. If the first digit is a 0 in the string, it won't
```
Do we actually need to do something here apart from checking that it is of the format `digit*.digit*`?
It is probably important to check that `1.0x2` is not a valid literal.

VoR0220 replied:
Yes...we do. If the first digits are 0s they cannot be converted into bigints.

chriseth commented on libsolidity/ast/Types.cpp:
```diff
@@ -353,16 +508,57 @@ bool IntegerConstantType::isValidLiteral(const Literal& _literal)
+	}
+	else
+	{
+		numerator = bigint(_literal.value());
```
`m_value = bigint(_literal.value())` does not work?

VoR0220 replied:
good point.

chriseth commented on libsolidity/ast/Types.cpp:
```diff
@@ -353,16 +508,57 @@ bool IntegerConstantType::isValidLiteral(const Literal& _literal)
+			char const* sub = Token::toString(_literal.token());
+			BOOST_THROW_EXCEPTION(
+				Error(Error::Type::TypeError) << errinfo_comment(
+					"Cannot divide sub denomination " + ((sub) ? string(Token::toString(_literal.token())) : "") + " any further."
```
If we add this check here, we should also forbid things like
`0.000000000000000000000000000001 szabo`. I think it is best not to check it at all.

VoR0220 replied:
Shouldn't we give some kind of warning for it in some form?

chriseth replied:
Why? `ufixed x = .5 wei` is perfectly fine, you will get an error when you want to convert it to `uint`.

VoR0220 replied:
ahhhhhhhh...okay I see your reasoning now.

chriseth commented on libsolidity/ast/Types.cpp:
```diff
@@ -390,10 +586,11 @@ IntegerConstantType::IntegerConstantType(Literal const& _literal)
-	if (auto targetType = dynamic_cast<IntegerType const*>(&_convertTo))
+	if (_convertTo.category() == Category::Integer)
 	{
+		auto targetType = dynamic_cast<IntegerType const*>(&_convertTo);
```
Is that needed?

chriseth commented on libsolidity/ast/Types.cpp:
```diff
@@ -406,28 +603,52 @@ bool IntegerConstantType::isImplicitlyConvertibleTo(Type const& _convertTo) cons
+		if (m_value == 0)
+			return true;
+		int forSignBit = (targetType->isSigned() ? 1 : 0);
+		if (m_value > 0)
```
What about negative values?

VoR0220 replied:
good point...this is a bit copy and paste from the integer portion.

chriseth commented on libsolidity/ast/Types.cpp:
```diff
@@ -406,28 +603,52 @@ bool IntegerConstantType::isImplicitlyConvertibleTo(Type const& _convertTo) cons
+		int forSignBit = (targetType->isSigned() ? 1 : 0);
+		if (m_value > 0)
+		{
+			bool properlyScaledBits = m_scalingFactor <= targetType->fractionalBits() ? 
```
It think it should be enough to test whether `m_value * pow(bigint(10), target_type->fractionalBits())` can be converted to the signed/unsigned integer type with `target_type->fractionalBits() + target_type->integerBits()` bits.

VoR0220 replied:
thinking about this...this should actually be checking to see whether the whole numbers of the rational can fit in the integer bits and if the remainder can fit in the fractional bits...Otherwise we haven't properly converted, have we?

chriseth commented on libsolidity/ast/Types.cpp:
```diff
@@ -485,73 +733,110 @@ TypePointer IntegerConstantType::binaryOperatorResult(Token::Value _operator, Ty
 	// we ignore the literal and hope that the type was correctly determined
-	solAssert(m_value <= u256(-1), "Integer constant too large.");
-	solAssert(m_value >= -(bigint(1) << 255), "Integer constant too small.");
+	solAssert(m_value <= u256(-1), "Number constant too large.");
```
I think this should be relaxed here to allow fractional literals.

VoR0220 replied:
what would you put instead?

chriseth commented on libsolidity/ast/Types.cpp:
```diff
@@ -485,73 +733,110 @@ TypePointer IntegerConstantType::binaryOperatorResult(Token::Value _operator, Ty
+shared_ptr<IntegerType const> ConstantNumberType::integerType() const
 {
-	bigint value = m_value;
+	bigint value = m_value.numerator();
```
Can we assert that the denominator is 1 here?

VoR0220 replied:
no, we don't want to do that. We need the conversion to integerType to take into account a case whereby a integer may need to be converted into a fixed point type and vice versa. Perhaps this is a clunky way of going about it, but that should be the case.

VoR0220 replied:
In fact I realized that the way this is going about it is all wrong. This needs to get rid of whatever remainder is left and truncate it.

chriseth commented on libsolidity/ast/Types.h:
```diff
@@ -338,9 +384,13 @@ class IntegerConstantType: public Type
 private:
-	bigint m_value;
+	rational m_value;
+	short m_scalingFactor;
```
I think we should not store the scaling factor, sometimes it does not even apply, for example for `1/3`.

VoR0220 replied:
that is what I was worried about. However it is very useful when converting to a fixed point type. I am currently running into problems surrounding this. What would you recommend instead of a scaling factor for a conversion?

chriseth replied:
So yes, the point where we have a problem is `fixed x = 1/3`, because it will never fit without truncating.
So the question is: What is the "intrinsic fixed point type" for a constant (what is returned by `fixedPointType()`) and which fixed point types do we allow implicit conversion to (explicit conversion is always fine)?

I would say the `canonical` fixed point type is the one with the smallest number of bits such that the number fits without losing information: I.e. we have to count the number of bits that can hold the numerator plus sign bit (should be the same code as in the integer case) and for the fractional bits, we should get that `denominator * 10**bits` is an integer. Then implicit conversions are ok if and only if the canonical fixed point type can be converted.

We should also add that information to the documentation and to https://github.com/ethereum/wiki/wiki/Solidity-Features

VoR0220 replied:
Well here's the problem. How does the function know the number of bits that it is converting to?

chriseth replied:
I'm confused, doesn't the description above explain that? `number of fractional bits = min{i | m_value * 2**i is integer and i is multiple of 8 and i <= 256}`

VoR0220 replied:
yes but remember, the NumberConstant and especially the fixedPoint function don't necessarily know that without an input.

VoR0220 replied:
wait a second...I might be wrong. Let me try something.

chriseth commented on libsolidity/formal/Why3Translator.cpp:
```diff
@@ -589,7 +589,7 @@ bool Why3Translator::visit(Literal const& _literal)
 		break;
-	case Type::Category::IntegerConstant:
+	case Type::Category::NumberConstant:
 		add("(of_int " + toString(type->literalValue(&_literal)) + ")");
```
Same here

chriseth commented on libsolidity/parsing/Token.cpp:
```diff
@@ -151,20 +153,19 @@ tuple<Token::Value, unsigned int, unsigned int> Token::fromIdentifierOrKeyword(s
 				positionX < _literal.end() &&
 				*positionX == 'x' &&
-				all_of(positionX + 1, _literal.end(), ::isdigit)
+				all_of(++positionX, _literal.end(), ::isdigit)
```
I made some changes here that are reverted now. Are you sure your changes are correct?

VoR0220 replied:
correct in the sense that they work. See below the reasoning...However....I can change it back to what it was before with little pain, so I think that can be done.

VoR0220 replied:
The changes for m and n needed to be put there, and the issue with UFixedMxN not being considered an identifier are also solved below.

VoR0220 replied:
there may be an odd error coming from my change. Investigating.

chriseth commented on libsolidity/ast/Types.cpp:
```diff
@@ -406,28 +603,52 @@ bool IntegerConstantType::isImplicitlyConvertibleTo(Type const& _convertTo) cons
 	}
+	else if (_convertTo.category() == Category::FixedPoint)
+	{
+		auto targetType = dynamic_cast<FixedPointType const*>(&_convertTo);
```
As discussed below, I think we should rather use `fixedPoint() && fixedPoint()->isImplicitlyConvertibleTo(_convertTo)` to avoid code duplication.

chriseth commented on libsolidity/ast/Types.cpp:
```diff
@@ -406,28 +603,52 @@ bool IntegerConstantType::isImplicitlyConvertibleTo(Type const& _convertTo) cons
+		TypePointer intType = integerType();
+		return intType && intType->isExplicitlyConvertibleTo(_convertTo);
+	}
+	TypePointer fixType = fixedPointType();
```
I think we have to do that more manually here, because `1/3` does not have a canonical fixed point type, but `fixed(1/3)` is still a valid explicit conversion.

chriseth replied:
If only integer types are concerned, we disallowed conversion when the number was too large.
Now that we add rationals, we cannot disallow conversion for rationals that cannot be represented in finite precision.
So the conversion rules are a bit more complicated and we have to check each of them here.

VoR0220 committed:
currently what we have

VoR0220 committed:
changed names for Rational Constants and categories

VoR0220 committed:
rational renaming

VoR0220 committed:
fixing modulus and Solidity Name and Type Resolution

minor fixes

current attempts at binary fixup

VoR0220 committed:
added bytes conversion tests, resolved that, converted to binary scaling, refactored the find algo to prevent large numbers and take into account integer bytes

think we're good on solidity type name resolution now

removed couts

updates to documentation and more removed couts along with literal value implementation

forgot semicolons

chriseth commented on libsolidity/ast/Types.cpp:
```diff
@@ -452,31 +639,57 @@ TypePointer IntegerConstantType::binaryOperatorResult(Token::Value _operator, Ty
 	{
-		bigint value;
+		rational value;
+		bool fixedPointType = (m_value.denominator() != 1 || other.m_value.denominator() != 1);
```
"`fractional`"?

VoR0220 replied:
Definitely

chriseth commented on libsolidity/ast/Types.cpp:
```diff
@@ -452,31 +639,57 @@ TypePointer IntegerConstantType::binaryOperatorResult(Token::Value _operator, Ty
 		case Token::BitOr:
-			value = m_value | other.m_value;
+			if (fixedPointType)
+				BOOST_THROW_EXCEPTION(Error(Error::Type::TypeError) << errinfo_comment("Cannot perform bit operations on non integer fixed type."));
```
These should return `TypePointer()`.

VoR0220 replied:
Fair enough.

chriseth commented on libsolidity/ast/Types.cpp:
```diff
@@ -486,72 +699,110 @@ TypePointer IntegerConstantType::binaryOperatorResult(Token::Value _operator, Ty
+			else if (fixedPointType)
+			{
+				value = m_value;
+				if (value > other.m_value)
```
Is that really what we want? If yes, is there a more efficient way to do it? :-)

VoR0220 replied:
Do tell me of the more efficient way to do it :) I've been thinking about it but I can't figure it out.

chriseth replied:
`a % b = c <=> 0 <= c < |b| && \exist x \in \int: a = xb+c`
So what about something like `a - ((a/b).numerator() / (a/b).denominator()) * b`?
Concerning the sign of `a%b`, please follow the convention used by the EVM itself.

chriseth commented on libsolidity/ast/Types.cpp:
```diff
@@ -486,72 +699,110 @@ TypePointer IntegerConstantType::binaryOperatorResult(Token::Value _operator, Ty
 		case Token::Mul:
 			value = m_value * other.m_value;
-			break;
+			break;			
```
Superfluous whitespace.

VoR0220 replied:
whoops

chriseth commented on libsolidity/ast/Types.h:
```diff
@@ -338,9 +384,15 @@ class IntegerConstantType: public Type
 
 	/// @returns the smallest integer type that can hold the value or an empty pointer if not possible.
 	std::shared_ptr<IntegerType const> integerType() const;
+	/// @returns the smallest fixed type that can hold the value or an empty pointer
```
Should be changed to "... hold the value or incurs the least precision loss. If the integer part does not fit, returns an empty pointer."

VoR0220 replied:
alright. I'll change it to that...and ensure that it actually does that.

chriseth commented on libsolidity/ast/Types.cpp:
```diff
@@ -561,6 +812,75 @@ shared_ptr<IntegerType const> IntegerConstantType::integerType() const
 
+shared_ptr<FixedPointType const> RationalNumberType::fixedPointType() const
+{
+	//do calculations up here
```
What about this implementation:

```
    bool negative = (m_value < 0);
    unsigned fractionalBits = 0;
    rational value = m_value;
    rational maxValue = rational(bigint(1) << 256);
    if (!negative)
        maxValue -= 1;
    else
        value = -value;
    while (value * 0x100 <= maxValue && value.denominator() != 1)
    {
        value *= 0x100;
        fractionalBits += 8;
    }

    if (value > maxValue)
        return shared_ptr<FixedPointType const>();
    bigint v = value.denominator() / value.numerator();
    if (negative)
        v = -v;
    // u256(v) is the actual value that will be put on the stack
    // From here on, very similar to integerType()
    if (negative) // convert to positive number of same bit requirements
        value = ((0 - value) - 1) << 1;
    if (value > u256(-1))
        return shared_ptr<FixedPointType const>();
    unsigned integerBits = bytesRequired(value);
    solAssert(integerBits >= fractionalBits, "Invalid bit requirement calculation.");
    //@todo special handling for integerBits == 0 && fractionalBits == 0?
    return make_shared<FixedPointType>(
        integerBits, fractionalBits,
        negative ? FixedPointType::Modifier::Signed : FixedPointType::Modifier::Unsigned
    );
```

VoR0220 replied:
Why would we solAssert that integerbits be bigger than fractionalBits? That takes away the ability to have a fixed0x256.

VoR0220 replied:
also, getting an error thrown at me with this:

rational maxValue = rational(bigint(1) << 256);

VoR0220 replied:
got it to run...but it's breaking most of my tests.

VoR0220 replied:
it's also somehow enabling 272 bit fraction types....

chriseth replied:
I did not say that it is perfect, but at least to me, it looks a lot simpler. I might be missing out on some edge cases, but actually I think I got all of them.

VoR0220 replied:
one I fixed it up it appears to be working...still gotta figure out why this infinite loop is happening though.

VoR0220 replied:
here's my patched up version. For some reason it is effecting my unary operation test with the negative...

```
shared_ptr<FixedPointType const> RationalNumberType::fixedPointType() const
{
    bool negative = (m_value < 0);
    unsigned fractionalBits = 0;
    unsigned integerBits = bytesRequired(wholeNumbers()) * 8;
    rational value = m_value;
    bigint l = bigint(1) << 256;
    rational maxValue = rational(l);
    if (!negative)
        maxValue -= 1;
    else
        value = -value;
    cout << "Max value: " << maxValue << endl;
    while (value * 0x100 <= maxValue && value.denominator() != 1 && fractionalBits < 256 - integerBits)
    {
        value *= 0x100;
        fractionalBits += 8;
    }

    if (value > maxValue)
    {
        cout << "value > max value " << endl;
        return shared_ptr<FixedPointType const>();
    }
    bigint v = value.denominator() / value.numerator();
    if (negative)
        v = -v;
    // u256(v) is the actual value that will be put on the stack
    // From here on, very similar to integerType()
    //if (negative) // convert to positive number of same bit requirements
    //  value = ((0 - value) - 1) << 1;
    if (value > u256(-1))
    {
        cout << "Too large of a number" << endl;
        return shared_ptr<FixedPointType const>();
    }

    //solAssert(integerBits >= fractionalBits, "Invalid bit requirement calculation.");
    //@todo special handling for integerBits == 0 && fractionalBits == 0?
    return make_shared<FixedPointType>(
        integerBits, fractionalBits,
        negative ? FixedPointType::Modifier::Signed : FixedPointType::Modifier::Unsigned
    );
}
```

chriseth commented on libsolidity/ast/Types.cpp:
```diff
@@ -390,10 +561,11 @@ IntegerConstantType::IntegerConstantType(Literal const& _literal)
-	if (auto targetType = dynamic_cast<IntegerType const*>(&_convertTo))
+	if (_convertTo.category() == Category::Integer)
 	{
+		auto targetType = dynamic_cast<IntegerType const*>(&_convertTo);
```
`If (m_value.denominator() != 1) return false;`

VoR0220 replied:
in this case though, there is an implicit conversion, we simply truncate off the fractional bits.

chriseth replied:
We cannot ever destroy information for an implicit conversion, that is one of its two core principles.

chriseth commented on libsolidity/ast/Types.cpp:
```diff
@@ -406,44 +578,59 @@ bool IntegerConstantType::isImplicitlyConvertibleTo(Type const& _convertTo) cons
 	}
+	else if (_convertTo.category() == Category::FixedPoint)
+	{
+		if (fixedPointType() && fixedPointType()->isImplicitlyConvertibleTo(_convertTo))
```
We cannot use `fixedPointType()` here because we have to return false if the number does not fit into finite precision. Instead, I would propose to do something like

```
rational shifted = (m_value * (bigint(1) << (8 * _convertTo.fractionalBits())));
if (shifted.denominator() != 1) return false;
// here, do the same stuff we do above for integers with _convertTo.integerBits() + _convertTo.fractioalBits() bits.
```

VoR0220 replied:
But it's already checking to see if the number can fit into finite precision, is it not?

VoR0220 replied:
see fixed point implicit conversion function.

chriseth commented on libsolidity/ast/Types.cpp:
```diff
@@ -406,44 +578,59 @@ bool IntegerConstantType::isImplicitlyConvertibleTo(Type const& _convertTo) cons
+		if (m_value.denominator() == 1)
+			return fixedBytes.numBytes() * 8 >= integerType()->numBits();
+		else
+			return fixedBytes.numBytes() * 8 >= fixedPointType()->numBits();
```
I would return false here because there is no obvious way to do the conversion.

VoR0220 replied:
Yes there is :) we talked about this. We take the fixed point bigint, we convert it over and take it in its pure byte form and pad right.

chriseth replied:
That assumes a certain encoding of the fixed number type. Remember that integer -> byte conversion is only explicit.

VoR0220 replied:
ahhhh...crap...yeah you're right.

VoR0220 committed:
much better way of doing modulus

VoR0220 committed:
changed algorithm for finding bits

VoR0220 committed:
updated algorithm for bit finding...now to figure out literal value

tiny fixups

changed location of the check

got rid of extra space and fixed a couple of things

added binary results bits

change back literal value

chriseth commented on libsolidity/analysis/ReferencesResolver.cpp:
```diff
@@ -103,10 +103,11 @@ void ReferencesResolver::endVisit(ArrayTypeName const& _typeName)
 		if (!lengthType)
 			fatalTypeError(length->location(), "Invalid array length.");
+		else if (lengthType->denominator() != 1)
+			fatalTypeError(length->location(), "Invalid input for array length, expected integer.");
```
Please change to `Invalid array length, expected integer literal.` (and please also change above)

chriseth commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -773,8 +773,8 @@ bool TypeChecker::visit(VariableDeclarationStatement const& _statement)
-				valueComponentType->category() == Type::Category::IntegerConstant &&
-				!dynamic_pointer_cast<IntegerConstantType const>(valueComponentType)->integerType()
+				valueComponentType->category() == Type::Category::RationalNumber &&
+				!dynamic_pointer_cast<RationalNumberType const>(valueComponentType)->integerType()
```
This should also work for rationals.

chriseth commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -799,8 +799,8 @@ bool TypeChecker::visit(VariableDeclarationStatement const& _statement)
-	if (type(_statement.expression())->category() == Type::Category::IntegerConstant)
-		if (!dynamic_pointer_cast<IntegerConstantType const>(type(_statement.expression()))->integerType())
+	if (type(_statement.expression())->category() == Type::Category::RationalNumber)
+		if (!dynamic_pointer_cast<RationalNumberType const>(type(_statement.expression()))->integerType())
```
Should also work for rationals.

chriseth commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -1106,7 +1106,7 @@ bool TypeChecker::visit(FunctionCall const& _functionCall)
 			{
-				if (auto t = dynamic_cast<IntegerConstantType const*>(argType.get()))
+				if (auto t = dynamic_cast<RationalNumberType const*>(argType.get()))
 					if (!t->integerType())
```
Should also work for rationals.

chriseth commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -1341,9 +1341,13 @@ bool TypeChecker::visit(IndexAccess const& _access)
-				if (!actualType.isDynamicallySized() && actualType.length() <= integerType->literalValue(nullptr))
+			if (auto numberType = dynamic_cast<RationalNumberType const*>(type(*index).get()))
+			{
+				if (numberType->denominator() != 1)
```
You can change that to an assert - we already checked for implicit convertibility to `u256` in line 1343.

chriseth commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -1368,7 +1372,7 @@ bool TypeChecker::visit(IndexAccess const& _access)
 		{
 			index->accept(*this);
-			if (auto length = dynamic_cast<IntegerConstantType const*>(type(*index).get()))
+			if (auto length = dynamic_cast<RationalNumberType const*>(type(*index).get()))
```
Please check that it is an integer here.

chriseth commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -1368,7 +1372,7 @@ bool TypeChecker::visit(IndexAccess const& _access)
 		else
 		{
 			index->accept(*this);
```
I think this should actually be `expectType(*index, IntegerType(256));`

chriseth commented on libsolidity/ast/Types.cpp:
```diff
@@ -251,17 +262,30 @@ IntegerType::IntegerType(int _bits, IntegerType::Modifier _modifier):
+		IntegerType const& convertTo = dynamic_cast<IntegerType const&>(_convertTo);
+		if (convertTo.m_bits < m_bits)
+			return false;
+		if (isAddress())
```
This restriction should also be checked when converting to rational

VoR0220 replied:
Rational or fixed point?

chriseth replied:
of course, fixed, conversion to rational is not possible.

chriseth commented on libsolidity/ast/Types.cpp:
```diff
@@ -307,7 +331,7 @@ string IntegerType::toString(bool) const
-	if (_other->category() != Category::IntegerConstant && _other->category() != category())
+	if (_other->category() != Category::RationalNumber && _other->category() != category())
 		return TypePointer();
 	auto commonType = dynamic_pointer_cast<IntegerType const>(Type::commonType(shared_from_this(), _other));
```
The conditions below are not 100% correct. It will allow the following:

```
uint8 x = 2;
ufixed y = .5;
x**y
```

Please make a note somewhere to add a test for that once we get to implementing fixed point types themselves.

VoR0220 replied:
I thought we were allowing that? That we would be able to run exponents of fixed types?

chriseth replied:
Right, I'm backing off :-)
But: be prepared to implement a gigantic mess of assembly for that.

VoR0220 replied:
YAY!!! :)

chriseth commented on libsolidity/ast/Types.cpp:
```diff
@@ -340,11 +364,133 @@ MemberList::MemberMap IntegerType::nativeMembers(ContractDefinition const*) cons
+	m_integerBits(_integerBits), m_fractionalBits(_fractionalBits), m_modifier(_modifier)
+{
+	solAssert(
+		m_integerBits + m_fractionalBits > 0 && 
```
Actually I would consider disallowing `fixedMx0` because that is just an alias for `intM` (this is not the place to do that, of course, but rather Token.cpp).

VoR0220 replied:
for me it would seem that it might make type conversion easier for the user.

chriseth replied:
Why do you think?

VoR0220 replied:
just simplicity of the idea and especially for explicit conversion. Be odd if we just out of the blue disallowed them from having a fixed act as an int...but I suppose it could go both ways. Me personally I'd like the flexibility in the understanding of it.

chriseth replied:
If we allow `fixedNx0`, then it should be an exact alias for `intN`, which I believe is more work to implement.

chriseth commented on libsolidity/ast/Types.cpp:
```diff
@@ -340,11 +364,133 @@ MemberList::MemberMap IntegerType::nativeMembers(ContractDefinition const*) cons
+		IntegerType const& convertTo = dynamic_cast<IntegerType const&>(_convertTo);
+		if (convertTo.numBits() < m_integerBits)
+			return false;
+		else if (isSigned())
```
Disallow conversion to address, please.

chriseth commented on libsolidity/ast/Types.cpp:
```diff
@@ -340,11 +364,133 @@ MemberList::MemberMap IntegerType::nativeMembers(ContractDefinition const*) cons
+	else if (_convertTo.category() == Category::Integer)
+	{
+		IntegerType const& convertTo = dynamic_cast<IntegerType const&>(_convertTo);
+		if (convertTo.numBits() < m_integerBits)
```
Conversion should only be allowed if `m_fractionalBits == 0`, but please see above (remove that case altogether).

chriseth commented on libsolidity/ast/Types.cpp:
```diff
@@ -340,11 +364,133 @@ MemberList::MemberMap IntegerType::nativeMembers(ContractDefinition const*) cons
+
+TypePointer FixedPointType::binaryOperatorResult(Token::Value _operator, TypePointer const& _other) const
+{
+	if (_other->category() != Category::RationalNumber 
```
```
if (
    _other->category()...
)
```

chriseth commented on libsolidity/ast/Types.cpp:
```diff
@@ -340,11 +364,133 @@ MemberList::MemberMap IntegerType::nativeMembers(ContractDefinition const*) cons
+TypePointer FixedPointType::binaryOperatorResult(Token::Value _operator, TypePointer const& _other) const
+{
+	if (_other->category() != Category::RationalNumber 
+		&& _other->category() != category()
```
Should be `||`

chriseth commented on libsolidity/ast/Types.cpp:
```diff
@@ -340,11 +364,133 @@ MemberList::MemberMap IntegerType::nativeMembers(ContractDefinition const*) cons
+		return commonType;
+	if (Token::isBooleanOp(_operator))
+		return TypePointer();
+	return commonType;
```
I think we should check all the operators, there might be some we have to disallow.

VoR0220 replied:
Bitop was one I just added.

chriseth commented on libsolidity/ast/Types.cpp:
```diff
@@ -340,11 +364,133 @@ MemberList::MemberMap IntegerType::nativeMembers(ContractDefinition const*) cons
+		auto radixPoint = find(_literal.value().begin(), _literal.value().end(), '.');		
+		if (radixPoint != _literal.value().end())
+		{
+			//problem here. If the first digit is a 0 in the string, it won't
```
Please change comment: Only decimal notation allowed here, leading zeros would switch to octal.

chriseth commented on libsolidity/ast/Types.cpp:
```diff
@@ -340,11 +364,133 @@ MemberList::MemberMap IntegerType::nativeMembers(ContractDefinition const*) cons
+		{
+			//problem here. If the first digit is a 0 in the string, it won't
+			//turn it into a integer...Using find if not to count the leading 0s.
+			auto leadingZeroes = find_if_not(
```
Please also check that the whole number conforms to `[0-9]*.[0-9]*`

VoR0220 replied:
can you elaborate?

chriseth commented on libsolidity/ast/Types.cpp:
```diff
@@ -353,15 +499,37 @@ bool IntegerConstantType::isValidLiteral(const Literal& _literal)
+		auto fractionalBegin = leadingZeroes != _literal.value().end() ?
+			leadingZeroes : radixPoint + 1;
+		//separatly grabbing the numerator, denominator for conversions
+		denominator = bigint(string(fractionalBegin, _literal.value().end()));
```
Code duplication with previous function.

VoR0220 replied:
indeed...should make this its own function.

chriseth commented on libsolidity/ast/Types.cpp:
```diff
@@ -486,72 +696,105 @@ TypePointer IntegerConstantType::binaryOperatorResult(Token::Value _operator, Ty
 		case Token::Mul:
 			value = m_value * other.m_value;
-			break;
+			break;	
```
Surplus whitespace.

chriseth commented on libsolidity/ast/Types.cpp:
```diff
@@ -561,6 +804,55 @@ shared_ptr<IntegerType const> IntegerConstantType::integerType() const
+		integerBits = bytesRequired(transitionValue) * 8;
+	}
+
+	while (value * 0x100 <= maxValue && value.denominator() != 1 && fractionalBits < 256 - integerBits)
```
I don't really get the meaning of `transitionValue` and I don't see why we need to keep track of `integerBits` here.
I would say this algorithm is much more stable if we first try to convert the number into an integer by shifting (as it is done here) and then define `integerBits = bytesRequired(value) - fractionalBits`.

VoR0220 replied:
hmmmm....that's a pretty clever way to go about it. I think we'll do that.

VoR0220 replied:
transitionValue is there because you cannot run bit operations on a rational type.

VoR0220 replied:
Also I put the integer bits in the while loop to make the loop more efficient. We go until we can't fit anymore and enables more numbers to get into the system that users would like to use. This is why we need it before we get the value.

VoR0220 replied:
relooking at it, I think we should keep it as is.

chriseth replied:
But then you have to add more comments, I don't get the meaning of `transitionValue`.

VoR0220 replied:
ahhhhh...it's a temporary big int to hold the value and pop it in the rational. Perhaps `temporaryBigInt` might be better?

VoR0220 replied:
just came to a realization...we shouldn't go ahead and loop through again on what is conceivably a larger number than we looped through for the fractional bits. If we only loop through for the integer bits, we're conceivably working with a smaller number of loops...there's considerable overlap with the v value of the same fractional bits that we just processed.

chriseth replied:
If there is an error in the implementation here, it might very well be that we will never notice.
Please compare "there might be an error we never notice" to "looping 20 or 32 times".

VoR0220 replied:
Hmmmm....i suppose when you frame it like that.

chriseth commented on libsolidity/ast/Types.h:
```diff
@@ -338,9 +384,15 @@ class IntegerConstantType: public Type
+	std::shared_ptr<FixedPointType const> fixedPointType() const;
+
+	bigint denominator() const { return m_value.denominator(); }
+	bigint wholeNumbers() const { return m_value.numerator() / m_value.denominator(); }
```
What about renaming this to `truncatedIntegerPart`?

VoR0220 replied:
Bit long of a name don't you think? What about something shorter?

chriseth replied:
integerPart?

VoR0220 replied:
that works

chriseth commented on libsolidity/ast/Types.cpp:
```diff
@@ -340,110 +367,254 @@ MemberList::MemberMap IntegerType::nativeMembers(ContractDefinition const*) cons
+				_literal.value().end(), 
+				[](char const& a) { return a == '0'; }
+			);
+			auto fractionalBegin = leadingZeroes != _literal.value().end() ?
```
It might be possible to remove this condition without replacement.

VoR0220 replied:
It's possible to remove this condition, but I noticed something very odd when I was testing out the var capabilities...only the first rational type registers...everything after that does not register atleast in terms of a isValid literal...it does not hit that function. This should require some investigation or atleast some explanation as to why this might be.

chriseth commented on libsolidity/ast/Types.cpp:
```diff
@@ -490,68 +687,100 @@ TypePointer IntegerConstantType::binaryOperatorResult(Token::Value _operator, Ty
+u256 RationalNumberType::literalValue(Literal const*) const
 {
 	u256 value;
+	bigint shiftedValue = integerPart();
```
We need `m_value * 2**numFixedBits` here. Also assert that denominator is 1 after multiplication. and then keep the additional asserts below. Oh and convert to 2's complement if negative.

VoR0220 replied:
how do we gather numFixedBits?

chriseth replied:
You don't need the actual number, it is fine to just multiply by 8 until the number is an integer.

VoR0220 replied:
if we're asking for the literal value...shouldn't we be getting the LITERAL value of the type? Shouldn't we be multiplying by 10 until denominator != 1?

innovator256 commented:
go go go guys....we are waiting :) thank you @VoR0220, thank you @chriseth !!

VoR0220 committed:
reorganized tests and fixed mobile types and implicit conversions of rationals and fixed point types

one final tweak

check for null types

chriseth commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -773,35 +773,56 @@ bool TypeChecker::visit(VariableDeclarationStatement const& _statement)
-				valueComponentType->category() == Type::Category::IntegerConstant &&
-				!dynamic_pointer_cast<IntegerConstantType const>(valueComponentType)->integerType()
+				valueComponentType->category() == Type::Category::RationalNumber &&
+				!dynamic_pointer_cast<RationalNumberType const>(valueComponentType)->integerType() &&
```
The call to `integerType` triggers an assertion if called on a non-integer. I guess the best thing here is to check that `mobileType()` does not return a null shared pointer.

This could be done in a generic way (not only for RationalNumber category) and then tune the error message in the case we have a RationalNumber.

chriseth commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -773,35 +773,56 @@ bool TypeChecker::visit(VariableDeclarationStatement const& _statement)
+						var.annotation().type->toString() +
+						". Try converting to type " +
+						valueComponentType->mobileType()->toString() +
+						" or using an explicit conversion." 
```
Wording: "or use"? Oh and `mobileType()` can return null!

chriseth committed:
Prefer mobileType() to check rational range.

chriseth committed:
Correctly handle unexpected exceptions during tests.

chriseth committed:
Replace "unsigned int" by "unsigned".

chriseth committed:
Some cleanup.

chriseth committed:
Disallow explicit fixed->enum, but allow bytes->fixed.

chriseth committed:
Cleanup.

chriseth committed:
Special case for moving sign bit to fractional part.

chriseth committed:
Simplify interface of RationalNumber.

chriseth committed:
Remove unused tests and add asserts for not implemented parts in code generation.

quick fix on christian's rational

change so that ubuntu will stop yelling

be more specific with rational declaration for Windows sake

rational in namespace correction for windows
